### PR TITLE
Downloaded manifests to temporary folders on AWS

### PR DESCRIPTION
**Pull request by linglp**
Related to the bug reported by @andrewelamb here: https://sagebionetworks.jira.com/browse/FDS-299

Highlights: 
* Added function in general.py that could create temporary folders in specified directory. 
* when downloading a manifest on AWS, downloading the manifests to a temporary location.
* Updated permission in Dockerfile to make sure that the user that we defined in uWSGI.ini could create folders on AWS. 

Will merge to develop after #1142

### CODE REVIEW

File: schematic/store/synapse.py [hunk 1]
```diff
@@ -65,18 +69,31 @@ class ManifestDownload(object):
...
     manifest_id: str
 
-    def _download_manifest_to_folder(self):
+    def _download_manifest_to_folder(self) -> File:
         """
         try downloading a manifest to local cache or a given folder
         manifest
         Return: 
-            manifest_data: A new Synapse Entity object of the appropriate type
+            manifest_data: synapse file entity
```
Comment (andrewelamb) [inline schematic/store/synapse.py:77]: The type should be in parenthesis:

manifest_data(File): synapse file entity

Perhaps the description could be more descriptive as well like "A Synapse file entity of the downloaded manifest"

File: schematic/store/synapse.py [hunk 2]
```diff
@@ -65,18 +69,31 @@ class ManifestDownload(object):
...
         """
-         # TO DO: potentially deprecate the if else statement because "manifest_folder" key always exist in config
-        if CONFIG["synapse"]["manifest_folder"]:
+        # TO DO: potentially deprecate the if else statement because "manifest_folder" key always exist in config
+        # on AWS, to avoid overriding manifest, we download the manifest to a temporary folder
+        if "SECRETS_MANAGER_SECRETS" in os.environ:
```
Comment (andrewelamb) [inline schematic/store/synapse.py:81]: This is a really complex series of if/elif/else statements, and really hard to parse. Is there any way to clean this up and make it easier to understand? If not I would recommend adding more documentation.

File: schematic/utils/general.py [hunk 3]
```diff
@@ -114,7 +116,21 @@ def entity_type_mapping(syn, entity_id):
...
+    """This function creates a temporary directory in the specified directory 
+    Args:
+        path(str): a directory path where all the temporary files will live
+    Returns: returns the absolute pathname of the new directory.
+    """
+    # Create a temporary directory in the specified directory
+    try:
+        path = tempfile.mkdtemp(dir=path)
+        return path
+    except FileNotFoundError:
```
Comment (andrewelamb) [inline schematic/utils/general.py:130]: Why is the same exception being raised after being caught?
If that is intended, I would [chain](https://docs.python.org/3/tutorial/errors.html#exception-chaining) this exception.

File: schematic/store/synapse.py [hunk 4]
```diff
@@ -65,18 +69,31 @@ class ManifestDownload(object):
...
         """
         try downloading a manifest to local cache or a given folder
         manifest
         Return: 
-            manifest_data: A new Synapse Entity object of the appropriate type
+            manifest_data: synapse file entity
         """
-         # TO DO: potentially deprecate the if else statement because "manifest_folder" key always exist in config
-        if CONFIG["synapse"]["manifest_folder"]:
+        # TO DO: potentially deprecate the if else statement because "manifest_folder" key always exist in config
```
Comment (andrewelamb) [inline schematic/store/synapse.py:79]: This todo should be removed and a jira ticket started instead.

**Comment by andrewelamb** [comment]
@linglp Should I check to see if I can recreate the bug via swagger with this branch?

File: schematic/store/synapse.py (hunk 2 above)
Comment (thomasyu888) [reply schematic/store/synapse.py:81]: +1. This chunk of code seems to be logic for `_determine_download_location`.  The main question I have is this: 

* Why does "SECRETS_MANAGER_SECRETS" existing in `os.environ` take precedence over `CONFIG['synapse']['manifest']`?

**Review by thomasyu888 — COMMENTED** [review]
Nice review by Andrew!  Just some minor comments on my end since you added me as reviewer.  Thanks for including me!

File: schematic/store/synapse.py (hunk 2 above)
Comment (linglp) [reply schematic/store/synapse.py:81]: @thomasyu888 That's because the config on AWS also has "manifest_folder" key. If I have `if CONFIG["synapse"]["manifest_folder"]` takes precedence, then the logic of downloading to a temporary folder can't be triggered. 

I think an alternative solution is to use a different config for deploying to AWS. Since in the future, we would use the develop branch to deploy to AWS, we can't add a different key in config.yml to indicate deployment. But maybe we could have a separate config.yml just for deployment? Will that bring in more complexity?

File: schematic/utils/general.py (hunk 3 above)
Comment (linglp) [reply schematic/utils/general.py:130]: Thanks for the feedbacks @andrewelamb and @thomasyu888 . I will update the code to remove try except.

File: schematic/store/synapse.py (hunk 1 above)
Comment (linglp) [reply schematic/store/synapse.py:77]: Thanks Andrew. Let me update the comment like you suggested.

File: schematic/store/synapse.py (hunk 2 above)
Comment (thomasyu888) [reply schematic/store/synapse.py:81]: Thanks - this is helpful. Similar to how AWS does credentials and profiles, maybe something similar can be done here? Can open a jira ticket to discuss.

I don't know enough of the codebase, but potentially you can see if there are duplicated keys in your CONFIG and look into simplifying functions to take in the the 'synapse' or 'aws' parameter instead of if-else statements.  You could have different "profile" configurations like you said when deploying differently.   I'll defer this decision to the FAIR team.

**Comment by linglp** [comment]
> @linglp Should I check to see if I can recreate the bug via swagger with this branch?

Hmm you could check if it works locally. But I don't think if you could check if that resolves the error on AWS by testing it locally.

**Comment by andrewelamb** [comment]
@linglp Yeah I guess we'll have to wait till its in AWS.

File: tests/test_utils.py [hunk 5]
```diff
@@ -77,6 +78,20 @@ def test_entity_type_mapping_invalid_entity_id(self, synapse_store):
         with pytest.raises(SynapseHTTPError) as exception_info:
             entity_type_mapping(syn, "syn123456")
 
+    def test_download_manifest_to_temp_folder(self):
+        # define temporary folder path
+        current_dir = os.getcwd()
```
[the remark this answers is not in the document]
Comment (linglp) [reply tests/test_utils.py]: Thanks Tom. This is a great suggestion.

File: schematic/store/synapse.py (hunk 2 above)
Comment (linglp) [reply schematic/store/synapse.py:81]: @andrewelamb I discussed this issue with Milen in our 1:1. He suggested that let's leave this code here for now and see if the solution works for fixing your bug. If it does, we could circle back to this issue and discuss with the team what would be a good way to handle downloading manifests on AWS v.s. downloading manifests locally

File: schematic/store/synapse.py [hunk 6]
```diff
@@ -65,18 +69,31 @@ class ManifestDownload(object):
...
+                os.mkdir("/var/tmp/temp_manifest_download")
+            download_location = create_temp_folder(temporary_manifest_storage)
+
+        elif CONFIG["synapse"]["manifest_folder"]:
+            download_location=CONFIG["synapse"]["manifest_folder"]
+
+        else:
+            download_location=None
+
+        if download_location:
```
[the remark this answers is not in the document]
Comment (linglp) [reply schematic/store/synapse.py]: Thanks Tom. I will take notes and update the code.

**Review by andrewelamb — APPROVED** [review]

**Comment by MiekoHash** [comment]
Hi @linglp @andrewelamb Is this something we can plan for sprint 15 to finish up verification on AWS?
